### PR TITLE
Promote admissionreview to v1

### LINKED ISSUE #79893: Promote AdmissionReview API to v1
Part of https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md

* Promote the API type to v1
* Teach the webhook dispatcher to send/receive v1
* Tighten validation of response objects when receiving v1 responses (require kind, apiVersion, status.uid)
* Allow "v1" in admissionReviewVersions for webhook configurations

/milestone v1.16
/priority important-soon
/sig api-machinery

### PR DESCRIPTION
**Pull request by liggitt**
**What this PR does / why we need it**:
* Promotes the AdmissionReview API types to v1 with no schema changes
* Updates the webhook dispatcher to be able to send/receive v1 AdmissionReview objects
* Tightens response validation for v1 AdmissionReview objects to verify apiVersion/kind/response.patchType/response.uid
* Adds integration tests to exercise registering for v1 webhooks receiving/returning v1 AdmissionReview objects

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubernetes/issues/79893

**Does this PR introduce a user-facing change?**:
```release-note
The `AdmissionReview` API sent to and received from admission webhooks has been promoted to `admission.k8s.io/v1`. Webhooks can specify a preference for receiving `v1` AdmissionReview objects with `admissionReviewVersions: ["v1","v1beta1"]`, and must respond with an API object in the same `apiVersion` they are sent. When webhooks use `admission.k8s.io/v1`, the following additional validation is performed on their responses:
* `response.patch` and `response.patchType` are not permitted from validating admission webhooks
* `apiVersion: "admission.k8s.io/v1"` is required
* `kind: "AdmissionReview"` is required
* `response.uid: "<value of request.uid>"` is required
* `response.patchType: "JSONPatch"` is required (if `response.patch` is set)
```

### CODE REVIEW

**Review by khenidak — APPROVED** [review]
/lgtm for `Fix panic in pod location serving` commit.

**Comment by liggitt** [comment]
This is ready for review now

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/request/admissionreview.go [hunk 1]
```diff
@@ -17,16 +17,183 @@ limitations under the License.
...
 )
 
-// CreateAdmissionReview creates an AdmissionReview for the provided admission.Attributes
-func CreateAdmissionReview(versionedAttributes *generic.VersionedAttributes, invocation *generic.WebhookInvocation) admissionv1beta1.AdmissionReview {
+// VerifyAdmissionResponse checks the validity of the provided admission review object, and returns the
+// audit annotations, whether the response allowed the request, any provided patch/patchType/status,
+// or an error if the provided admission review was not valid.
+func VerifyAdmissionResponse(uid types.UID, mutating bool, review runtime.Object) (auditAnnotations map[string]string, allowed bool, patch []byte, patchType string, result *metav1.Status, err error) {
+	switch r := review.(type) {
+	case *admissionv1.AdmissionReview:
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/request/admissionreview.go]: Thanks for adding the response validation for v1. This looks good.

File: pkg/apis/admission/install/install.go [hunk 2]
```diff
@@ -34,5 +35,6 @@ func init() {
 func Install(scheme *runtime.Scheme) {
 	utilruntime.Must(admission.AddToScheme(scheme))
 	utilruntime.Must(v1beta1.AddToScheme(scheme))
-	utilruntime.Must(scheme.SetVersionPriority(v1beta1.SchemeGroupVersion))
+	utilruntime.Must(v1.AddToScheme(scheme))
+	utilruntime.Must(scheme.SetVersionPriority(v1beta1.SchemeGroupVersion, v1beta1.SchemeGroupVersion))
```
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/apis/admission/install/install.go]: good eye

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go [hunk 3]
```diff
@@ -216,14 +212,20 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *v1beta
...
 	jsonSerializer := json.NewSerializer(json.DefaultMetaFactory, o.GetObjectCreater(), o.GetObjectTyper(), false)
-	objJS, err := runtime.Encode(jsonSerializer, attr.VersionedObject)
-	if err != nil {
-		return false, apierrors.NewInternalError(err)
-	}
-	patchedJS, err := patchObj.Apply(objJS)
-	if err != nil {
-		return false, apierrors.NewInternalError(err)
+	switch patchType {
+	case string(admissionv1.PatchTypeJSONPatch):
```
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: will change to the version-independent PatchTypeJSONPatch constant

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/request/admissionreview.go [hunk 4]
```diff
@@ -17,16 +17,183 @@ limitations under the License.
...
+		if r.Response == nil {
+			return nil, false, nil, "", nil, fmt.Errorf("webhook response was absent")
```
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/request/admissionreview.go]: but nil, false, nil, "", nil has such a nice ring to it :)

**Comment by liggitt** [comment]
comments addressed

**Comment by liggitt** [comment]
/hold
for API ack from @smarterclayton

**Comment by smarterclayton** [comment]
API LGTM 

/approve
/lgtm

Unhold at your leisure

File: staging/src/k8s.io/api/admission/v1/types.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+	// Resource is the fully-qualified resource being requested (for example, v1.pods)
+	Resource metav1.GroupVersionResource `json:"resource" protobuf:"bytes,3,opt,name=resource"`
+	// SubResource is the subresource being requested, if any (for example, "status" or "scale")
+	// +optional
+	SubResource string `json:"subResource,omitempty" protobuf:"bytes,4,opt,name=subResource"`
+
+	// RequestKind is the fully-qualified type of the original API request (for example, v1.Pod or autoscaling.v1.Scale).
+	// If this is specified and differs from the value in "kind", an equivalent match and conversion was performed.
+	//
+	// For example, if deployments can be modified via apps/v1 and apps/v1, and a webhook registered a rule of
```
Comment (roycaihw) [inline staging/src/k8s.io/api/admission/v1/types.go]: via apps/v1 and apps/v1beta1

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/request/admissionreview_test.go [hunk 6]
```diff
@@ -0,0 +1,618 @@
...
+import (
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
```
Comment (roycaihw) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/request/admissionreview_test.go]: could you remind me why we are still using v1beta1 internally? Do we have plan to switch the configuration manager and the webhook packages to use v1?

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/request/admissionreview_test.go]: Yes, we will. Just wanted to keep this one focused on the admissionreview API

File: staging/src/k8s.io/api/admission/v1/types.go (hunk 5 above)
Comment (liggitt) [reply staging/src/k8s.io/api/admission/v1/types.go]: Will update in a follow up

File: pkg/apis/admissionregistration/validation/validation.go [hunk 7]
```diff
@@ -152,6 +152,10 @@ func validateRule(rule *admissionregistration.Rule, fldPath *field.Path, allowSu
 	return allErrors
 }
 
+// AcceptedAdmissionReviewVersions contains the list of AdmissionReview versions the *prior* version of the API server understands.
```
Comment (roycaihw) [inline pkg/apis/admissionregistration/validation/validation.go]: this is to make sure we can safely downgrade apiserver one version? I was surprised that you cannot create a configuration that only talks v1 in 1.16

Comment (liggitt) [reply pkg/apis/admissionregistration/validation/validation.go]: to be able to safely downgrade, and to ensure a skewed HA cluster of 1.15 and 1.16 API servers does not permit creation of webhooks that will break half the API servers in the cluster

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/request/admissionreview_test.go (hunk 6 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/request/admissionreview_test.go]: opened https://github.com/kubernetes/kubernetes/pull/80883 to do so, though I think we'll hold that for 1.17

File: staging/src/k8s.io/api/admission/v1/types.go (hunk 5 above)
Comment (liggitt) [reply staging/src/k8s.io/api/admission/v1/types.go]: fixed in https://github.com/kubernetes/kubernetes/pull/80884

**Pull request merged**